cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitor.java

```diff
@@ -225,74 +296,124 @@ protected void run() {
+          store.deleteSubscriberState(topicConf, programRunId.getRun());
+        }
+      }), retryStrategy);
     }
   }
 
+  private Map<String, MonitorConsumeRequest> initTopics() {
+    Map<String, MonitorConsumeRequest> consumeRequests = new HashMap<>();
+
+    Transactionals.execute(transactional, context -> {
+      AppMetadataStore store = AppMetadataStore.create(cConf, context, dsFramework);
+
+      for (String topicConfig : requestKeyToLocalTopic.keySet()) {
+        String messageId = store.retrieveSubscriberState(topicConfig, programRunId.getRun());
+        consumeRequests.put(topicConfig, new MonitorConsumeRequest(messageId, limit));
+      }
+    });
+
+    return consumeRequests;
+  }
+
   /**
    * Processes the given set of monitoring messages.
    *
    * @param monitorResponses set of messages to be processed.
    * @return the latest message publish time or {@code -1L} if there is no message to process
    */
-  private long processResponse(Map<String, List<MonitorMessage>> monitorResponses) throws Exception {
+  private long processResponse(Map<String, List<MonitorMessage>> monitorResponses,
+                               AppMetadataStore store) throws Exception {
     long latestPublishTime = -1L;
 
     for (Map.Entry<String, List<MonitorMessage>> monitorResponse : monitorResponses.entrySet()) {
       String topicConfig = monitorResponse.getKey();
       List<MonitorMessage> monitorMessages = monitorResponse.getValue();
 
-      if (monitorMessages.isEmpty()) {
+      // Skip terminal program status messages to delay the deprovisioning.
+      if (monitorMessages.isEmpty() ||
+              (topicConfig.equals(Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC) && programFinishTime > 0)) {
         continue;
       }
 
-      if (topicConfig.equals(Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC)) {
-        if (programFinishTime < 0) {
-          programFinishTime = findProgramFinishTime(monitorMessages);
-        }
+      long publishTime = publish(topicConfig, requestKeyToLocalTopic.get(topicConfig), monitorMessages, store);
+      latestPublishTime = Math.max(publishTime, latestPublishTime);
+    }
 
-        if (programFinishTime >= 0) {
-          // Buffer the program state messages and don't publish them until the end
-          // Otherwise, once we publish, the deprovisioner will kicks in and deleting the cluster
-          // which could result in losing the last set of messages for some topics.
-          lastProgramStateMessages.addAll(monitorMessages);
-          // We still update the in memory store for the next fetch offset to avoid fetching duplicate
-          // messages, however, that shouldn't be persisted to avoid potential loss of messages
-          // in case of failure
-          updateTopicToRequest(topicConfig, monitorMessages);
-          continue;
-        }
+    return latestPublishTime;
+  }
+
+  private void updateConsumeRequests(Map<String, List<MonitorMessage>> monitorResponses) {
```

Comment:
It'll be good to have some basic comments for non-obvious method even if it is private (no comments for getter/setter are fine)